### PR TITLE
Introduce `NodeTemplate` attribute in MachineClass

### PR DESCRIPTION
**Pull request by AxiomSamarth**
**What this PR does / why we need it**:
This PR is needed to implement the generic scale to zero feature in CA. This PR adopts the `cpu`, `gpu` and the `memory` attributes of the `expectedNodeAttribute` to be of type `resource.Quantity`.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
The release of MCM including this change will have to be vendored in CA. So, we would want to prioritise this one.

**Release note**:

```feature user
NONE
```

### CODE REVIEW

File: pkg/apis/machine/v1alpha1/machineclass_types.go [hunk 1]
```diff
@@ -73,20 +74,17 @@ type MachineClassList struct {
 // ExpectedNodeAttributes contains subfields to track all scale from zero resources
 type ExpectedNodeAttributes struct {
 	// CPU cores of the expected node
-	// +optional
-	CPU *int64 `json:"cpu,omitempty"`
+	CPU resource.Quantity `json:"cpu,omitempty"`
```
Comment (himanshu-kun) [inline pkg/apis/machine/v1alpha1/machineclass_types.go:77]: we can remove `omitempty` from required fields

Comment (AxiomSamarth) [reply pkg/apis/machine/v1alpha1/machineclass_types.go:77]: Yes

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
Thanks for the quick changes Samarth
* we should add a release note of change in the API maybe 
* there's not enough validation for the fields , but we have discussed that and it wouldn't be solved until we do validation by webhook
* pls let me know your comments if `instanceType` should also be made required
* Currently `region` and `zone` are part of `ExpectedNodeAttributes` but in CA they can be fetched from other places in machineClass as well. I don't remember the reason why we included them in expectedNodeAttributes. Can you explain that also.

**Review by himanshu-kun — APPROVED** [review]
/lgtm

File: pkg/apis/machine/v1alpha1/machineclass_types.go [hunk 2]
```diff
@@ -39,18 +40,17 @@ type MachineClass struct {
 	// +optional
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
+	// +kubebuilder:validation:XPreserveUnknownFields
+	// AutoscalerNodeTemplate contains subfields to track all node resources required to scale nodegroup from zero
+	AutoscalerNodeTemplate AutoscalerNodeTemplate `json:"autoscalerNodeTemplate"`
```
Comment (himanshu-kun) [inline pkg/apis/machine/v1alpha1/machineclass_types.go:45]: we might have to make this field `optional` (and so pointer as well) for backward compatibility as stated for the `NodeTemplate` field by Tim [here](https://github.com/gardener/gardener/pull/4980#discussion_r754852941)

Comment (AxiomSamarth) [reply pkg/apis/machine/v1alpha1/machineclass_types.go:45]: Yes, I had already made a point about it.

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Comment by AxiomSamarth** [comment]
> /lgtm

Even though this attribute is an optional attribute but the fields when mentioned are pretty critical. So, I am thinking to add a validation for the MachineClass field [here](https://github.com/gardener/machine-controller-manager/blob/f6c94b820534e7aeaf12e99f1b9fe59eea77ceb0/pkg/util/provider/machinecontroller/machine_util.go#L96) for the map `Capacity` to have the keys `cpu`, `gpu` and `memory` which are critical right now.

**Comment by himanshu-kun** [comment]
> Even though this attribute is an optional attribute but the fields when mentioned are pretty critical. So, I am thinking to add a validation for the MachineClass field here for the map Capacity to have the keys cpu, gpu and memory which are critical right now.

Yes we can do that. However with this validation we won't be able to stop CA to read the machineClass. 
From what I understand,  this validation can help is when CA instructs to scale a machineDeployment from zero, and if cpu,gpu or memory not present , then machine won't scale up.
So yeah in one way it'll work.

**Comment by AxiomSamarth** [comment]
Yes, that the idea. The scaleup of nodes belonging to a nodeGroup that is consuming a faulty machine class should fail and the node group should go into unhealthy state. As and when the machineClass validations are moved as a Webhook this behaviour will improve.

**Comment by AxiomSamarth** [comment]
But anyway, a check will have to be added in the CA also for these attributes or else, there could be a panic.

File: pkg/util/provider/machinecontroller/machine_util.go [hunk 3]
```diff
@@ -133,6 +133,11 @@ func (c *controller) ValidateMachineClass(ctx context.Context, classSpec *v1alph
 		return nil, nil, machineutils.ShortRetry, err
 	}
 
+	err = c.validateNodeTemplate(machineClass.NodeTemplate)
+	if err != nil {
+		return nil, nil, machineutils.ShortRetry, err
```
Comment (himanshu-kun) [inline pkg/util/provider/machinecontroller/machine_util.go:138]: before return we can add an error log that node template is not valid.

File: pkg/util/provider/machinecontroller/machine_util.go [hunk 4]
```diff
@@ -158,6 +163,25 @@ func (c *controller) getSecretData(machineClassName string, secretRefs ...*v1.Se
...
+		return nil
+	}
+
+	for _, attribute := range capacityAttributes {
+		if _, ok := nodeTemplate.Capacity[attribute]; !ok {
+			err := errors.New("MachineClass NodeTemplate Capacity should mandatorily have CPU, GPU and Memory configured")
+			return err
+		}
+	}
+
```
Comment (himanshu-kun) [inline pkg/util/provider/machinecontroller/machine_util.go:181]: we can add validation for instanceType , region and zone , validating that empty string is not provided

Comment (himanshu-kun) [reply pkg/util/provider/machinecontroller/machine_util.go:181]: the docstring of the method will have to be adopted accordingly

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
looks good, minor changes requested

File: pkg/util/provider/machinecontroller/machine_util.go (hunk 3 above)
Comment (AxiomSamarth) [reply pkg/util/provider/machinecontroller/machine_util.go:138]: good point.

File: pkg/util/provider/machinecontroller/machine_util.go (hunk 4 above)
Comment (AxiomSamarth) [reply pkg/util/provider/machinecontroller/machine_util.go:181]: good point too.

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Pull request merged**